singularity-eos/eos/modifiers/ramps_eos.hpp

```diff
@@ -0,0 +1,213 @@
+  return;
+}
+
+template <typename T>
+class SAPRampEOS : public EosBase<SAPRampEOS<T>> {
+ public:
+  // move semantics ensures dynamic memory comes along for the ride
+  SAPRampEOS(T &&t, const Real r0, const Real a, const Real b, const Real c)
+      : t_(std::forward<T>(t)), r0_(r0), a_(a), b_(b), c_(c),
+        rmid_(r0 * (a - b * c) / (a - b)) {
+    // add input parameter checks to ensure validity of the ramp
+    assert(r0 >= 0.0);
+    assert(a > 0.0);
+    assert(b >= 0);
+    assert(a != b);
+  }
+  SAPRampEOS() = default;
+
+  auto GetOnDevice() { return SAPRampEOS<T>(t_.GetOnDevice(), r0_, a_, b_, c_); }
+  inline void Finalize() { t_.Finalize(); }
+
+  PORTABLE_INLINE_FUNCTION
+  Real get_ramp_pressure(Real rho) const {
+    const Real p_ramp{rho < r0_     ? 0.0
+                      : rho < rmid_ ? a_ * (rho / r0_ - 1.0)
+                                    : b_ * (rho / r0_ - c_)};
+    return p_ramp;
+  }
+
+  PORTABLE_FUNCTION
+  Real TemperatureFromDensityInternalEnergy(const Real rho, const Real sie,
+                                            Real *lambda = nullptr) const {
+    return t_.TemperatureFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real InternalEnergyFromDensityTemperature(const Real rho, const Real temperature,
+                                            Real *lambda = nullptr) const {
+    return t_.InternalEnergyFromDensityTemperature(rho, temperature, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real PressureFromDensityInternalEnergy(const Real rho, const Real sie,
+                                         Real *lambda = nullptr) const {
+    // ramp pressure
+    const Real p_ramp{get_ramp_pressure(rho)};
+    // pressure from eos
+    const Real p_eos{t_.PressureFromDensityInternalEnergy(rho, sie, lambda)};
+    // return max(p_ramp, p_eos)
+    return p_eos < p_ramp ? p_ramp : p_eos;
+  }
+  PORTABLE_FUNCTION
+  Real SpecificHeatFromDensityInternalEnergy(const Real rho, const Real sie,
+                                             Real *lambda = nullptr) const {
+    return t_.SpecificHeatFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real BulkModulusFromDensityInternalEnergy(const Real rho, const Real sie,
+                                            Real *lambda = nullptr) const {
+    return t_.BulkModulusFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real GruneisenParamFromDensityInternalEnergy(const Real rho, const Real sie,
+                                               Real *lambda = nullptr) const {
+    return t_.GruneisenParamFromDensityInternalEnergy(rho, sie, lambda);
+  }
+  PORTABLE_FUNCTION
+  Real PressureFromDensityTemperature(const Real rho, const Real temperature,
+                                      Real *lambda = nullptr) const {
+    const Real p_ramp{get_ramp_pressure(rho)};
+    const Real p_eos{t_.PressureFromDensityTemperature(rho, temperature, lambda)};
```

Comment:
for my own edification, is there a particular reason to use curly bracket initialization here rather than just the assignment operator?

Reply:
Yes, like I said (even though I'm inconsistent with it), it is a good habit to have. Though for POD types it's not needed.

Reply:
I don't think in this context it matters much... Despite being around @dholladay00 for a long time now, I still haven't picked up this habit for scalars. :P